### PR TITLE
[SPARK-9968][STREAMING] Reduced time spent within synchronized block to prevent lock starvation

### PR DESCRIPTION
**Pull request by tdas**
When the rate limiter is actually limiting the rate at which data is inserted into the buffer, the synchronized block of BlockGenerator.addData stays blocked for long time. This causes the thread switching the buffer and generating blocks (synchronized with addData) to starve and not generate blocks for seconds. The correct solution is to not block on the rate limiter within the synchronized block for adding data to the buffer.

### CODE REVIEW

**Comment by tdas** [comment]
@zsxwing Could you take a look?

File: streaming/src/main/scala/org/apache/spark/streaming/receiver/BlockGenerator.scala [hunk 1]
```diff
@@ -106,6 +106,7 @@ private[streaming] class BlockGenerator(
   private val blockQueueSize = conf.getInt("spark.streaming.blockQueueSize", 10)
   private val blocksForPushing = new ArrayBlockingQueue[Block](blockQueueSize)
   private val blockPushingThread = new Thread() { override def run() { keepPushingBlocks() } }
+  private val dataAddingLock = new Object
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/BlockGenerator.scala]: `dataAddingLock` is not used.

**Comment by zsxwing** [comment]
LGTM except one minor comment

File: streaming/src/main/scala/org/apache/spark/streaming/receiver/BlockGenerator.scala (hunk 1 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/BlockGenerator.scala]: Right, sorry, I thought of using it but I realized that it changes the locking logic significantly, which I dont want to do.

**Comment by tdas** [comment]
Thanks @zsxwing. I removed the unused variable. Since the tests already passed,  I will merge this as soon as this compiles.

**Comment by tdas** [comment]
Merging this master and branch 1.5.

**Pull request closed without merging**